### ISSUE TITLE
Missing poll event when using io_uring_register

### ISSUE THREAD

**Issue by santigimeno** [axboe/liburing#214]
Hi,

I've a scenario where a parent process communicates with a child process through a `socketpair` using `sendmsg/recvmsg`. I'm polling the fds of the `socketpair` using `IORING_OP_POLL_ADD` so I can know when there's available data for reading. I'm also registering those fd's. They send data back and forth and after a while the parent process closes the connection using `close()`. The issue I'm facing is that the child process doesn't detect the disconnection (there's no `poll` event received). The problem looks similar to when trying to close a connection without having unregistered the fd first, but I'm doing it just before calling `close()` by using `io_uring_register_files_update()` with -1 . If I run the same code without registering the files, it works correctly and the disconnection is detected.

I'm trying to come up with a reduced test case that demonstrates the issue as I can only reproduce it as part of a larger test case but in the meantime any hint on what I could be doing wrong would really be appreciated.

Thanks in advance.

**Comment by santigimeno** [comment]
Hi again,

I've been looking into this and have a much simpler test case that reproduces the issue: https://gist.github.com/santigimeno/daa8342778a81031ebf81a49ca0b41a4. 

It's a single process with a pair of sockets sending data back and forth and then one of the sockets closes the connection and when the other end detects it, it closes itself and the process exits. I'm registering both sockets and polling using `IORING_OP_POLL_ADD`. If no other requests are sent to the ring everything works fine and the disconnection is detected. On the other hand if I register and poll some other unrelated fd which doesn't receive any event, the disconnection is not detected  (in the example is the read end of a pipe). Then, if I close this pipe, the disconnection is detected again. 

Probably I'm doing something wrong but cannot figure out what. Any help would be appreciated. Thanks!

**Comment by isilence** [comment]
> I've been looking into this and have a much simpler test case that reproduces the issue: https://gist.github.com/santigimeno/daa8342778a81031ebf81a49ca0b41a4.

Reproducers are very helpful, thanks!
 
> It's a single process with a pair of sockets sending data back and forth and then one of the sockets closes the connection and when the other end detects it, it closes itself and the process exits. I'm registering both sockets and polling using `IORING_OP_POLL_ADD`. If no other requests are sent to the ring everything works fine and the disconnection is detected. On the other hand if I register and poll some other unrelated fd which doesn't receive any event, the disconnection is not detected (in the example is the read end of a pipe). Then, if I close this pipe, the disconnection is detected again.
> 
> Probably I'm doing something wrong but cannot figure out what. Any help would be appreciated. Thanks!

I'll take a look.

**Comment by axboe** [comment]
If you register a file, it holds a reference to it. That means that if you then close() it in your application, the final reference isn't released and it's effectively not really closed. It's akin to having done dup() on it in that regard, and only closing the original.

**Comment by isilence** [comment]
From the other hand, it's claimed the one that is closed is also unregistered
beforehand. see the description.

**Comment by santigimeno** [comment]
Yes, I'm unregistering the fd before calling `close()`.

**Comment by isilence** [comment]
@axboe, any chance a socketpair points to the same `struct file`?

**Comment by axboe** [comment]
Gotcha, thanks for clarifying!

socketpair() should be 2 separate files, even internally.

**Comment by isilence** [comment]
@axboe, got it, thanks.

As said, I'll look what's going on.

**Comment by isilence** [comment]
The example can be simplified to
```C
void test(int conn0, int conn1)
{
	register_fds(conn0, conn1);

	add_fixed_file_poll(conn1); // or any other fixed-file op
	unregister_fd(conn0);
	close(conn0);
}
```
In short, you see this bug because unregister() doesn't put a reference
to a struct file immediately but delays it until all* previous
fixed-file requests complete, and that creates a circular dependency.

TL;DR;
The fixed-file poll is not cancelled until `conn0` is dead, but it also
holds a ref to `struct fixed_file_ref_node` preventing from going away
and releasing a struct file reference to `conn0`

Finer-grained refcounting should help (e.g. per-file), but atomic
refcounting would make the whole thing pointless, and pcpu
doesn't seem good memory-wise. I'd prefer to find something
more elaborate.

**Comment by isilence** [comment]
Even though I don't like such circular dependencies, I don't
really see how to efficiently circumvent this.

@santigimeno, how about adding shutdown() then? It should
solve the issue and feels like the right way of doing that.

**Comment by santigimeno** [comment]
@isilence first of all, thanks for looking into it.
What do you mean with adding `shutdown()`? Do you refer to replacing `close()` with `shutdown()`?  If that's the case a couple of questions:
- Why does it solve the issue?
- How, a consumer of liburing knows when to use or not to use `close()` to avoid this problem? Also, why using `close()` may not be 'right'? 

Thanks in advance.

**Comment by isilence** [comment]
> @isilence first of all, thanks for looking into it.
> What do you mean with adding `shutdown()`? Do you refer to replacing `close()` with `shutdown()`? If that's the case a couple of questions:

@santigimeno, I mean prepending close() with shutdown() (+unregister if needed)
```C
shutdown(fd);
close(fd);
```



>     * Why does it solve the issue?

The problem is that a request (e.g. poll_add()) for `conn1` implicitly holds a reference
to `conn0`, and `conn0` won't actually close until the request is gone. It helps because
shutdown() doesn't care about number of reference to a file, but close() actually closes
when all dup()'ed fd's and other references are gone.

`shutdown(conn0)` makes `conn1` to react (i.e. complete the poll req) before refcounting
of `conn0` is actually downed to zero. So, shutdown() completes `poll(conn1)`, which
releases a ref to `conn0`, that allows it to be really closed.

>     * How, a consumer of liburing knows when to use or not to use `close()` to avoid this problem? Also, why using `close()` may not be 'right'?

Maybe "don't expect registered files to be _actually_ closed until io_uring is
destroyed"? That sounds to restrictive, though. It "... won't be close, until
all requests submitted prior to that close are completed", but it would better
to be thought through.

**Comment by YoSTEALTH** [comment]
@santigimeno for socket calling `shutdown` before `close` is a common practice. https://man7.org/linux/man-pages/man3/shutdown.3p.html

**Comment by santigimeno** [comment]
@YoSTEALTH yes, I know about that. My point is, calling `close()` directly is also a valid practice and knowing whether this is going to block a `fd` to be actually closed can be problematic from the API user POV.

**Comment by YoSTEALTH** [comment]
@santigimeno you do have a valid point, sure `close()` should be able to be called directly.

I also get what @isilence is saying about how reference keeps the `fd` open.

I suppose a better solution where user/technical point of view is addressed.
